kad-dht/README.md

```diff
@@ -0,0 +1,395 @@
+# libp2p Kademlia DHT specification
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and mainlineDHT \[3\].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Authors
+
+* Protocol Labs.
+
+## Editors
+
+* [Raúl Kripalani](https://github.com/raulk)
+* [John Hiesey](https://github.com/jhiesey)
+
+## Distance function (dXOR)
+
+The libp2p Kad DHT uses the **XOR distance metric** as defined in the original
+Kademlia paper [0]. Peer IDs are normalised through the SHA256 hash function.
+
+For recap, `dXOR(sha256(id1), sha256(id2))` is the number of common leftmost
+bits between SHA256 of each peer IDs. The `dXOR` between us and a peer X
```

Comment:
Hi. I'm trying to write a bit of doc about the DHT here: https://github.com/ipfs/docs/pull/240 
I still need to wrap my head around some things.

But shouldn't dXOR be "256 minus the number of common leftmost bits"?
The spec implies dXOR(id1,id1) = 256.  I would expect dXOR(id1,id1) = 0 an input should have a distance of 0 with itself.
The spec implies dXOR(0000001111...,00000000000.......) = 6.  I would expect dXOR(0000001111...,00000000000.......) = 250, as a big distance would imply a big difference.
Am I missing something here?

Reply:
This spec is currently incorrect. The actual distance is just the one in the kademlia paper: `id1 xor id2`. This section is confusing that with the how a peer's bucket is calculated. _That_ uses the number of shared bits:

```
Bucket 0: no shared bits.
Bucket 1: 1 shared bit.
Bucket 2: 2 shared bits.
...
Last bucket: everything else
```

Reply:
(fixed)